js_modules/dagster-ui/packages/ui-core/src/util/hashObject.ts

```diff
@@ -0,0 +1,188 @@
+          // Use a consistent string representation for numbers
+          hashChunk(current.value.toString());
+        } else if (typeof current.value === 'string') {
+          hashChunk(current.value);
+        }
+        current.state = 2; // Mark as done
+      } else if (current.type === TYPE_ARRAY) {
+        hashChunk(smallBuffer[2]); // '['
+
+        if (current.value.length === 0) {
+          hashChunk(smallBuffer[3]); // ']'
+          current.state = 2; // Mark as done
+        }
+      } else if (current.type === TYPE_OBJECT) {
+        hashChunk(smallBuffer[0]); // '{'
+
+        // Sort keys once and cache them
+        current.keys = Object.keys(current.value).sort();
+
+        if (current.keys.length === 0) {
+          hashChunk(smallBuffer[1]); // '}'
+          current.state = 2; // Mark as done
+        }
+      }
+    }
+    // Process array/object elements
+    else if (current.state === 1) {
+      if (current.type === TYPE_ARRAY) {
+        const arr = current.value;
+
+        if (current.index > 0) {
+          hashChunk(smallBuffer[5]); // ','
+        }
+
+        if (current.index < arr.length) {
+          const item = arr[current.index++];
+          const itemType =
+            item === null || typeof item !== 'object'
+              ? TYPE_PRIMITIVE
+              : Array.isArray(item)
+                ? TYPE_ARRAY
+                : TYPE_OBJECT;
+
+          // Push the item onto the stack
+          stack.push({
+            type: itemType,
+            value: item,
+            index: 0,
+            state: 0,
+          });
+        } else {
+          // Finished processing array
+          hashChunk(smallBuffer[3]); // ']'
+          current.state = 2;
+        }
+      } else if (current.type === TYPE_OBJECT) {
+        const keys = current.keys!;
+
+        if (current.index > 0) {
+          hashChunk(smallBuffer[5]); // ','
+        }
+
+        if (current.index < keys.length) {
+          const key = keys[current.index++]!;
+          hashChunk(key);
+          hashChunk(smallBuffer[4]); // ':'
+
+          const item = current.value[key];
+          const itemType =
+            item === null || typeof item !== 'object'
+              ? TYPE_PRIMITIVE
+              : Array.isArray(item)
+                ? TYPE_ARRAY
+                : TYPE_OBJECT;
+
+          // Push the item onto the stack
+          stack.push({
+            type: itemType,
+            value: item,
+            index: 0,
+            state: 0,
+          });
+        } else {
+          // Finished processing object
+          hashChunk(smallBuffer[1]); // '}'
+          current.state = 2;
+        }
+      }
+    }
+    // Finished with this item
+    else {
+      stack.pop();
+    }
+  }
+
+  // Finalize the hash (MurmurHash3 finalization)
+  h1 ^= h1 >>> 16;
+  h1 = (h1 * 0x85ebca6b) >>> 0;
+  h1 ^= h1 >>> 13;
+  h1 = (h1 * 0xc2b2ae35) >>> 0;
+  h1 ^= h1 >>> 16;
+
+  return h1.toString(16).padStart(8, '0');
```

Comment:
Hmm I'm surprised that this is faster than calling through to the C code for JSON.stringify, this is awesome!

Reply:
`JSON.stringify`  is faster but results in the full string needing to be allocated in memory which is the issue. With this approach we never have a 100MB+ string in memory.

Reply:
Ahh gotcha - I guess the other option would be to do a digest of some sort of the JSON string? I'm not sure the extent to which WebCrypto hashing is available I guess!

Reply:
We were doing that before but the fact that the whole json gets serialized as a string is the issue. With this approach we update a digest as we process the JSON rather than creating the whole string at once.